File: 11. Container With Most Water.java
//OPTIMISED SOLUTION - complexity - nlogn

class Solution {
    public int maxArea(int[] height) {
        
        int cor = 0;
        int area = 0;
        int leftpointer = 0;
        int rightpoointer = height.length-1;

        while(leftpointer!=rightpoointer)
        {
            cor = Math.min(height[leftpointer],height[rightpoointer]);
            area = Math.max(area ,cor * (rightpoointer - leftpointer)) ;
            //System.out.println("area: "+area + " rightpoointer:"+ rightpoointer +" leftpointer:"+leftpointer + " cor:"+cor);
            if(height[leftpointer]> height[rightpoointer])
            {
                rightpoointer = rightpoointer-1;
            }
            else
            {
                leftpointer = leftpointer+1;
            }
        }
        return area;

    }
}


//my approach-

class Solution {
    public int maxArea(int[] height) {

        int l = 0;
        int r = height.length-1;
        int res = Integer.MIN_VALUE;

        while(l<r)
        {
            int minheight = Math.min(height[l],height[r]);
            res = Math.max(res, area(l,r,minheight));
            if(height[l]<height[r])
            {
                l++;
            }
            else
            {
                r--;
            }
            
        }

        return res;
        
    }

    public static int area(int l, int r, int minheight)
    {
        return (r-l)*minheight;
    }
}

File: python_codes/11. Container With Most Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        p1 = 0
        p2 = len(height)-1
        maxarea = float('-inf')

        while(p1<p2):
        
            distance = p2-p1
            area = distance * min(height[p2],height[p1])
            maxarea = max(maxarea, area)
            #print("maxarea is", maxarea)

            if(height[p1]<height[p2]):
                p1+=1
            else:
                p2-=1
        
        return maxarea